### PR TITLE
[LoRA] Support PEFT trainable token rows alongside LoRA adapters

### PR DESCRIPTION
**Pull request by zupengwang**
PEFT adapters trained with `LoraConfig(trainable_token_indices=...)` save compact `token_adapter.trainable_tokens_delta` tensors. The current loader rejects those tensors. This change loads their absolute replacement rows and applies them to each request's input embeddings and selected output logits, including output-head weights omitted by PEFT for tied embeddings.

Users opt in with `--max-lora-trainable-tokens N` / `LLM(max_lora_trainable_tokens=N)`. Storage scales with selected rows and adapter slots. The default of zero allocates no replacement buffers. Ordinary LoRA on other layers can be used together with selected-token rows, and adapter eviction/reload clears the previous rows before slot reuse. Checkpoint/configuration errors are rejected before evicting or replacing an active adapter.

Related issue: https://github.com/vllm-project/vllm/issues/36501. This implements the generic PEFT capability for models that already expose embedding/head LoRA modules, such as Llama. It does not claim to complete Qwen3-VL embedding-module registration; that is separate work in https://github.com/vllm-project/vllm/pull/33234.

## Duplicate-work check

On 2026-09-08, the issue remained open and unassigned. Searches of open PRs for the issue number, `trainable_tokens`, `trainable_token_indices`, and `trainable token` found no competing generic implementation. Prior PR https://github.com/vllm-project/vllm/pull/36648 is closed without merging. This implementation uses absolute replacement rows and includes tied output-head semantics, compact runtime buffers, mixed-adapter kernels, and model parity coverage.

## Scope

- PEFT list and module-dictionary token configuration, module name mapping, safetensors and PyTorch checkpoint loading.
- Per-adapter compact embedding and output-logit replacement, tied weights, fp16/bf16/fp32 storage, and CUDA Graph replay.
- Model-specific shape, vocabulary, capacity, configuration, and overlap validation before cache mutation.
- Native embedding/head LoRA modules only; no vocabulary resizing, pipeline parallelism, or ordinary LoRA on the same embedding/head as selected-token replacements.

## Validation

Base: `54da70c1eb6976fe655efd535f14a01378cd08bc`.

- Focused loader, configuration, CPU/CUDA operations, graph replay, manager/worker validation and slot reuse: **80 passed, 25 deselected**.
- Real PEFT adapter parity on a locally generated two-layer Llama: **6 passed** across eager, compiled/CUDA Graph, and TP=2, each with tied and untied embeddings. Each case checks 42 requests, three generated tokens per request, exact greedy tokens and all 64 vocabulary logprobs against Transformers + PEFT. Fixed prompts have a reference top-two margin greater than 0.03; logprob tolerances are `atol=0.015`, `rtol=0.003`.
- Ruff check/format, mypy for Python 3.10 and 3.12, typos, Markdown lint, SPDX, root lazy imports, filename, forbidden import, torch CUDA API, configuration, boolean-context-manager, and applicable Docker graph checks passed.
- `git diff --check` passed; all 17 local source-file hashes match the files tested remotely.
- `torch.library.opcheck` passed schema, autograd-registration, FakeTensor and dynamic AOT checks for both new operators on CPU.

Commands (from an installed checkout):

```bash
.venv/bin/python -m pytest tests/lora/test_utils.py tests/lora/test_peft_helper.py tests/lora/test_trainable_tokens.py tests/lora/test_lora_manager.py -k 'trainable or test_parse_fine_tuned_lora_name or test_replace_submodule or test_get_adapter_absolute_path or test_peft_helper_invalid_rank_direct or test_bfloat16_replacement or test_embedding_rows or test_logits_match or test_padding_adapter or test_empty_batch or test_graph_replay' -q
.venv/bin/python -m pytest tests/lora/test_trainable_tokens_model.py -k 'eager or graph' -q
CUDA_VISIBLE_DEVICES=0,1 .venv/bin/python -m pytest tests/lora/test_trainable_tokens_model.py -k tp2 -q
```

The test host used RTX 3090 GPUs, PyTorch 2.13.0+cu130, Transformers 5.16.1 and PEFT 0.19.1. Python and new Triton kernels came from this exact checkout; unchanged native libraries were reused from an existing vLLM 0.28 environment. TP=2 initially exposed a missing upstream `vocab_parallel_embedding` op; that op was compiled from this base SHA and loaded as a scoped test-environment supplement. This is not a complete native rebuild of the base SHA. Full upstream CI, a matched full build, and a trained-model/task-accuracy evaluation have not run. The synthetic model tests establish numerical integration correctness, not task accuracy or a throughput improvement.

The all-hooks pre-commit invocation was interrupted during an unrelated actionlint Go dependency download; all hooks applicable to these changed files were subsequently invoked individually and passed.

## AI assistance and human accountability

Codex assisted with candidate research, implementation, regression tests and validation. The submitting human reviewed the change and authorized publication.

### CODE REVIEW

**Comment by linitra24** [comment]
CC @jeejeelee Is this feature useful?

**Comment by zupengwang** [comment]
@jeejeelee could you review the PEFT semantics here, especially absolute token-row replacement, tied embeddings/output heads, and adapter slot reuse? This enables serving PEFT `trainable_token_indices` adapters that the current loader rejects.

@hmellor could you also review the `--max-lora-trainable-tokens` option?

The conflicts with main are resolved in c09f21d. On that commit, 86 focused tests and 6 synthetic Llama/PEFT parity cases passed across eager execution, CUDA Graphs, and TP=2. Upstream pre-commit, DCO, and docs checks are green. The GPU tests reused existing native libraries; a full native rebuild remains outstanding, as detailed in the PR.

Thanks!
